rocket-cookie-auth/src/user.rs

```diff
@@ -0,0 +1,76 @@
+use crate::auth::rand_string;
+use crate::error::Error;
+use anyhow::Result;
+use argon2::verify_encoded;
+use rocket::http::Status;
+use rocket::request::FromRequest;
+use rocket::request::Outcome;
+use rocket::Request;
+use serde::Deserialize;
+use serde::Serialize;
+
+/// A request guard that can be included in handler definitions to enforce authentication.
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone, Hash, PartialOrd, Ord)]
+pub struct User {
+    pub id: u32,
+    #[serde(skip_serializing)]
+    pub(crate) password: String,
+    pub auth_key: String,
+    pub expired: bool,
+}
+
+#[rocket::async_trait]
+impl<'r> FromRequest<'r> for User {
+    type Error = Error;
+    async fn from_request(request: &'r Request<'_>) -> Outcome<User, Self::Error> {
+        use rocket::outcome::Outcome::*;
+        let guard = request.guard().await;
+        tracing::info!("Guard: {guard}");
+        let auth: crate::auth::Auth = match guard {
+            Success(auth) => auth,
+            Failure(x) => return Failure(x),
+            Forward(x) => return Forward(x),
+        };
+        if let Ok(Some(user)) = auth.get_user().await {
+            Success(user)
+        } else {
+            Failure((Status::Unauthorized, Error::Unauthorized))
+        }
+    }
+}
+
+impl User {
+    /// Change users password
+    pub fn set_password(&mut self, new: &str) -> Result<()> {
+        let password = create_password(new)?;
+        self.password = password;
+        Ok(())
+    }
+}
+
+pub fn create_password(password: &str) -> Result<String> {
+    let password = password.as_bytes();
+    let salt = rand_string(10);
+    let config = argon2::Config::default();
+    let hash = argon2::hash_encoded(password, salt.as_bytes(), &config)?;
```

Comment:
This should maybe be run inside a tokio `spawn_blocking` or `block_in_place` task

Reply:
Can you explain why?

Reply:
The process of encoding a hash is a CPU bound task, so it should not be run as-is on the runtime thread. I've used argon2 for a similar purpose before (though for validating session authentication tokens), and verifying them can take up to 11ms. See this example:
![image](https://user-images.githubusercontent.com/6688948/189881492-7c756c06-6186-4f0d-b21f-9ce25b796039.png)

Reply:
I failed using these function due to do some lifetime issues. :/

Reply:
What were the issues? Might need to clone or move in the password